### ISSUE TITLE
Error in model training with patches as float vs unsigned integer

### ISSUE THREAD

**Issue by Pratyush1991** [remicres/otbtf#25]
Hello, 

I had an NDVI composite of 26-time intervals. The image had pixels with nan values and -9999 was used for no data. So I used otbcli_ManageNoData to create a mask of valid and non-valid pixels. Then I used the otbcli_DynamicConvert to linear rescale the values between (-1, 1 ) to (0, 1). The mask was also used in the process. The output was a tif file of floating type (32bit). Then I followed the usual patch extraction process (otbcli_PatchExtraction) for training and validation data. I have three classes for classification. 
So I found that when I used a floating type image for model. The model showed errors like all the pixels were classified into one class.

![image](https://user-images.githubusercontent.com/28738918/86157391-8822e000-bb31-11ea-9547-28d4eca98df8.png)

So I used otbcli_DynamicConvert to linear rescale the values between 0,10000. In this case, I choose output as an unsigned integer type (16 bit). The model training was successful in this case. 

![image](https://user-images.githubusercontent.com/28738918/86157559-c7e9c780-bb31-11ea-93e1-d89e0b2f093f.png)

As I read in most of the cases that image should be normalized between 0 and 1 for training. So I am not sure if this is correct as well what might be causing the issue.

**Comment by remicres** [comment]
Hi @Pratyush1991 , 

1. can you put your model snippet here? What is the placeholders data type?

2. Having values in the range [-1, 1] is okay (rescaling in the [0, 1] range isn't necessary).

3. I don't know if you are using otbtf1.7 or later: in recent versions there is a parameter to reject the patches containing at leat one no-data value in `PatchesExtraction`: it can be useful to avoid corrupted patches.

4. As I understand, when you use the NDVI values that are in the [-1, 1] range, you have a bad training result. And when you scale those values between [0, 10000], you have correct result. My guess is that there is a problem somewhere with the pixel values. Did you check that the tif files were not writen in `int16` of something other that `float32`? for each output of an OTB application, you can change the pixel encoding.

**Comment by Pratyush1991** [comment]
Hello @remicres ,  

1.  The model that I used was simple CNN. I did not change anything except for placeholder value to 26. The datatype of placeholder was float32 (x) and int16 (y).

2. Noted. 

3. I was using this on otbtf1.7:gpu. As you mentioned I updated it to otbtf1.8 to use the parameter to reject the samples.

4. As you mentioned, I tried inspecting and found that the problem was with pixels having NaN values. Although I had -9999 set as no data in metadata still there were pixels with NaN value. I used NumPy to check the values. And when I converted the float to an unsigned integer, the NaN values are converted to 0 automatically. (As NaN is only for float type I think). I used the otbcli_manage_nodata to include NaN as no data and assign no data as -9999 for the float32 bit file. So in the new file which was float32 but all the NaN were set was -9999 which was missing in the previous float32 bit file. 

Then I did the patch extraction and rejected -9999 using the parameter for no data value. The model training worked  :)

Thank you for your help

Regards
Pratyush
